src/main/java/eu/europeana/normalization/normalizers/PidNormalizer.java

```diff
@@ -0,0 +1,159 @@
+
+    // Go by each proxy.
+    final List<ProxyType> proxies = Optional.ofNullable(rdf.getProxyList()).stream()
+        .flatMap(Collection::stream).toList();
+    final InternalNormalizationReport report = new InternalNormalizationReport();
+    for (ProxyType proxy : proxies) {
+
+      // Set up some collections: split into PIDs that need normalization and those that don't.
+      final List<Pid> allPidsInProxy = Optional.ofNullable(proxy.getPidList()).stream()
+          .flatMap(Collection::stream).filter(Objects::nonNull).toList();
+      final List<Pid> nonNormalizedPidsInProxy = allPidsInProxy.stream()
+          .filter(pid -> !StringUtils.isBlank(pid.getString())).toList();
+      final Set<String> normalizedPidIdsInProxy = allPidsInProxy.stream()
+          .map(Pid::getResource).filter(Objects::nonNull).map(Resource::getResource)
+          .filter(resource -> !StringUtils.isBlank(resource)).collect(Collectors.toSet());
+
+      // Normalize PIDs.
+      final List<Pid> resultPidsInProxy = new ArrayList<>();
+      for (Pid nonNormalizedPid : nonNormalizedPidsInProxy) {
+
+        // Normalize the PID. If we can't, add the PID directly as a result.
+        final PidMatchResult normalization = pidSchemeMatcher.apply(nonNormalizedPid.getString());
+        if (normalization == null) {
+          resultPidsInProxy.add(nonNormalizedPid);
+          continue;
+        }
+
+        // Find or create the normalized PID object. Collect IDs in a set to guarantee uniqueness.
+        final PersistentIdentifierType normalizedPid = findOrAddNormalizedPidIfNeeded(
+            normalization, normalizedPidsById);
+        normalizedPidIdsInProxy.add(normalizedPid.getAbout());
+
+        // Add to the report
+        report.increment(this.getClass().getSimpleName(), ConfidenceLevel.CERTAIN);
+      }
+
+      // Add the normalized PID IDs to the result list and replace the list in the proxy.
+      for (String id : normalizedPidIdsInProxy) {
+        final Pid newPid = new Pid();
+        newPid.setResource(new Resource());
+        newPid.getResource().setResource(id);
+        newPid.setString("");
+        resultPidsInProxy.add(newPid);
+      }
+      proxy.setPidList(resultPidsInProxy);
+    }
+
+    // Override all the normalized PIDs and PID schemes in the record as new ones were added.
+    rdf.setPersistentIdentifierList(new ArrayList<>(normalizedPidsById.values()));
+
+    // Done
+    return new NormalizeActionResult(RecordWrapper.create(rdf), report);
+  }
+
+  private static PersistentIdentifierType findOrAddNormalizedPidIfNeeded(
+      PidMatchResult normalization, Map<String, PersistentIdentifierType> normalizedPidsById) {
+
+    // Try to find a matching PID object.
+    final Optional<PersistentIdentifierType> existingPid = normalizedPidsById.values().stream()
+        .filter(candidate -> candidate.getValue().getString().equals(normalization.canonicalPid()))
```

Comment:
Can getValue or getString be null? Perhaps protect it here or earlier in map construction.